ValidationTweaker.php

```diff
@@ -645,54 +581,98 @@ protected function outputSurveyValidationSkip( $instrument )
+		if ( ! $hasDefaultCalc && ! $hasRandomnumber )
 		{
 			return;
 		}
-		$listActionTagsJS = [];
-		foreach ( $listActionTags as $t => $d )
+		$listFieldNames = \REDCap::getFieldNames( $instrument );
+		foreach ( $listFieldNames as $fieldName )
 		{
-			$listActionTagsJS[] = [ $t, $d ];
+			$annotation = $GLOBALS['Proj']->metadata[$fieldName]['misc'];
+			$annotation = \Form::replaceIfActionTag( $annotation, $this->getProjectId(), $record,
+			                                         $eventID, $instrument, $instance );
+			$defaultCalc = $hasDefaultCalc
+			               ? \Form::getValueInParenthesesActionTag( $annotation, '@DEFAULT-CALC' )
+			               : '';
+			if ( $defaultCalc != '' )
+			{
+				$defaultVal = \REDCap::evaluateLogic( $defaultCalc, $this->getProjectId(),
+				                                      $record, $eventID, $instance,
+				                                      $instrument, $instrument, null, true );
+				$defaultVal = ( strpos( $defaultVal, "'" ) === false )
+				              ? ( "'" . $defaultVal . "'" )
+				              : ( '"' . str_replace( '"', '', $defaultVal ) . '"' );
+				$GLOBALS['Proj']->metadata[$fieldName]['misc'] =
+						"@DEFAULT=" . $defaultVal . " " . $annotation;
+			}
+			elseif ( $hasRandomnumber &&
+			         preg_match( '/(^|\\s)@RANDOMNUMBER(\\((-?[0-9]+,-?[0-9]+)\\))?(\\s|$)/',
+			                     $annotation, $randomnumberMatches ) )
+			{
+				$randomnumber = 0;
+				$randomnumberRange = $randomnumberMatches[3] ?? '';
+				if ( $randomnumberRange == '' )
+				{
+					$randomnumber = random_int( 0, ( PHP_INT_MAX - 1 ) ) / PHP_INT_MAX;
+					$decimalPlaces = -1;
+					$validationType =
+							$GLOBALS['Proj']->metadata[$fieldName]['element_validation_type'];
+					if ( $validationType == 'integer' || $validationType == 'int' )
+					{
+						$decimalPlaces = 0;
+					}
+					elseif ( strpos( $validationType, 'number_' ) === 0 )
+					{
+						$decimalPlaces = intval( substr( $validationType, 7, 1 ) );
+					}
+					if ( $decimalPlaces > -1 )
+					{
+						$randomnumber = round( $randomnumber, $decimalPlaces );
+					}
+					if ( strpos( $validationType, '_comma_decimal' ) !== false )
+					{
+						$randomnumber = str_replace( '.', ',', $randomnumber );
+					}
+				}
+				else
+				{
+					$randomnumberRange = explode( ',', $randomnumberRange );
+					if ( $randomnumberRange[0] > $randomnumberRange[1] )
+					{
+						$randomnumberTemp = $randomnumberRange[1];
+						$randomnumberRange[1] = $randomnumberRange[0];
+						$randomnumberRange[0] = $randomnumberTemp;
+					}
+					$randomnumber = random_int( intval( $randomnumberRange[0] ),
+					                            intval( $randomnumberRange[1] ) );
+				}
+				$GLOBALS['Proj']->metadata[$fieldName]['misc'] =
+						"@DEFAULT='" . $randomnumber . "' " . $annotation;
+			}
 		}
-		$listActionTagsJS = json_encode( $listActionTagsJS );
+	}
+
+
+
+
+
+	// Output JavaScript to remove action tags from the action tags guide.
```

Comment:
Is it hidden or remove?

Reply:
Any optional action tags which have not been enabled in the module system settings are not displayed in the action tags guide.